---
title: "SlideSeqDNA_Vesalius_Dev"
output: html_document
---


## Settings - Parameters etc

Specify dataset + Normalisation + dimensionality reduction method and other parameters

For DNA use log and PCA, or use TFIDF and LSI?

```{r}
start.time <- Sys.time()
# SlideSeq DNA datasets:
#alias <- "human_colon_cancer_4_dna_200114_13"
#alias <- "human_colon_cancer_3_dna_191204_19"
alias <- "mouse_liver_met_2_dna_200114_10"



norm <- "TFIDF"             # Normalization method: "log","SCT","TFIDF","raw"
meth <- "LSI"               # Dimensionality reduction method: "PCA","PCA_L","UMAP","LSI","LSI_UMAP"
PCs <- 6                    # Number of Embeddings / dimensions / Principal Components / "LSI topics"
plotPC <- 6                 # Number of PC plots to export.

min.count <- 0
max.count <- 6

k <- 5
knn.type <- "Duplicate"     # KNN method: "Unique", "Duplicate"

cores <- 5

colD <- 6                   # Column Depth (3-6)
smIter <- 10                # Smoothing Iterations
#imSegDims <- 4             # Image Segmentation Dimensions
st.meth <- "wilcox"          #Statistical method for extracting markers/bins, "wilcox", "t.test", "chisq"


param <- paste(paste0(norm, "Norm"), paste0(meth, PCs), sep = "_")

in.path <- "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/"
out.path <- "~/Thesis_SpatialGenomics/R_Output/"
out.pathDev <- "~/Thesis_SpatialGenomics/R_Output/VesaliusDev/"


```

In paper they do k 50 for smoothing the beads before PCA (mouse liver)

## Load packages needed

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/")
knitr::opts_chunk$set(echo = TRUE)
library(imagerExtra)
library(Seurat)
library(ggplot2)
library(patchwork)
library(viridis)
library(RColorBrewer)
library(dplyr)
library(tvR)
library(Signac)
library(sp)
library(spdep)
library(grid)
library(gridExtra)
library(Matrix)
library(MatrixExtra)


files <- list.files("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R",pattern = ".R")
files <- paste0("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R/",files)
for(i in seq_along(files)){
  source(files[i])
}

# Standard libraries - just in case you are a command line fiend
library(utils)
library(stats)
library(graphics)
library(grDevices)
### Set seed
set.seed(1)
```



## Read data from files
Vectors:
barcodes    -     holds the barcode sequences
sparse      -     Sparse count matrix, dgCMatrix

```{r}
counts <- paste0(in.path, alias, ".sparse_counts_1Mb.txt")
bead <- ReadSlideSeq(paste0(in.path, alias, ".bead_locations.csv"))

barcodes <- rownames(bead@coordinates)

sep = "\t" #separator used in this count matrix
ctmp <- read.table(counts, header = TRUE, sep=sep ) %>% na.exclude()

sparse <- sparseMatrix(i= ctmp[,2],j=ctmp[,1],x=ctmp[,3]) 

# Set point size for plotting the UMIs - adjusted for knn later
pt.sz <- 0.25
```


## Genomic Bins Human/Mouse:
Each bin index corresponds to 1 million bases in a specific location on a chromosome.
bin size: 1 million bases, index 1 = chromosome 1 from base 1 -> 1.000.000
Keep only the bin index numbers (1-3114. ~3.1bn bases)

Human: 3114 bins human bins. Chr2 starts at bin 251
Mouse: 2738 bins total 

```{r}

if(grepl("human",alias)){
  bins <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/hg19_1Mb_bins.txt", header = T)
  bins <- bins$bin_ind
} else if(grepl("mouse",alias)){
  bins <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/mm10_1Mb_bins.txt", header = T)
  bins <- bins$bin_ind
}
```



## Name column/rows in sparse matrix
Assign bin index numbers to identify the rows. 
Assign the barcodes to identify the columns

```{r}
colnames(sparse) <- barcodes
rownames(sparse) <- bins
```

#Remove outlier counts
Expected CNA varies between 0-6? (Total deletion and double duplication?)
High outlier counts might represent mitochondrial DNA bins (probably not since I only have chromosomes in my bins list?) or repeated elements or other errors?


```{r}

if(exists("min.count")){
  sparse <- filterSparse(sparse, function(x) x >= min.count)
}
if(exists("max.count")){
  sparse <- filterSparse(sparse, function(x) x <= max.count)
  miax <- paste0("counts", min.count, "-", max.count)
  param <- paste(miax, param, sep="_")
}


```


## Vector holders - Checkpoint

```{r}
df1 <- as.matrix(sparse) 
bead1 <- bead
param1 <- param

df <- df1
bead <- bead1
param <- param1
```


## Run KNN
- Decide between only using each barcode once (KNN.DNA.Unique function)
  or having many duplicate groups of barcodes, each barcode gets a group of k nearest neighbors (KNN.DNA function)


```{r}
if(exists("k")){
  if(knn.type == "Unique"){
    ls <- KNN.DNA.Unique(bead, df, k)
  }else{
    ls <- KNN.DNA(bead, df, k)
  }
  param <- paste(param, paste0("knn", knn.type, k), sep="_")
  sparse <- ls[[1]] #Retrieve sparse matrix output from knn function
  bead <- ls[[2]] #Retrieve Bead info (barcodes and coordinates) in SlideSeq assay format compatible with Seurat/Vesalius object

  pt.sz <- k/15 #Adjusting point size for plotting
  rm(ls)
}

bead <- cbind(rownames(bead), bead[[1]], bead[[2]])
colnames(bead) <- c("barcodes","xcoord","ycoord")
bead <- as.data.frame(bead)
bead <- transform(bead, xcoord = as.numeric(xcoord), 
               ycoord = as.numeric(ycoord))

sparse.df <- as.data.frame(sparse)
sparse.spmx <- Matrix(sparse, sparse = TRUE)

```

## Create Vesalius Object - Milestone
Based on the counts and bead coordinates

Beads in data.frame format - coordinates numerical
Counts either dataframe or sparse matrix.

# Dimensionality Reduction
Latent space
PCA / UMAP / tSNE (/LSI)

set filterThreshold to remove outlier beads (0.99)
```{r}
#sparse <- sparse.df
sparse <- sparse.spmx

vesalius <- buildVesaliusObject(bead,sparse)
ves <- buildVesaliusEmbeddings(vesalius, method = meth, 
  norm = norm, pcs = PCs, tensorResolution = 0.1, cores = cores, filterThreshold =0.99)

```

## Plotting vesalius images of the PC or LSI embeddings

Imageplot with chosen number of dimensions for the different PCs

```{r}
cat( paste(Sys.time()," Plotting top", plotPC, "PCs and exporting to pdf.\n File located in:", out.pathDev, "\n"))
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "Embeddings", alias, paste0(param, ".pdf"), sep="_")
pdf(fname)
#layout(matrix(c(1,3,2,4), 2, 2))
for(i in seq(plotPC)){
  print(imagePlot(ves, dims = i))
}
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.pathDev, fname))

print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))
```


## Image processing:
imageSegmentation - iterative k-means clustering
  - dims: number of dimred embeddings
  - Embedding "last": last dimred method used (if you have multiple dimred obejcts stored in your ves object)
LayerTerritory - not relevant for DNA data?

  

```{r}
vesImSeg <- imageSegmentation(ves,dims=1:PCs, colDepth = colD, smoothIter=smIter,smoothType="iso",sigma=2,embedding=meth)

vesTerr <- isolateTerritories(vesImSeg, trial = "last")

param <- paste(param, paste0("dims1-", PCs), paste0("colD", colD), paste0("Iter", smIter), sep="_")
print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))
```



# Export territory plots to pdf
Is there a difference between the territory isolation and the territory morphing?
Also plot the first imagesegmentation

```{r}
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "Territories", alias, paste0(param, ".pdf"), sep="_")
pdf(fname)
layout(matrix(c(1,3,2,4), 2, 2))
territoryPlot(vesTerr,cex=10,cex.pt=pt.sz)
print(imagePlot(vesImSeg))
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.pathDev, fname))
```

## Extract marker bins and morph territories
ExtractMarkers - DGE between selected territories - identify marker bins
  - Takes like 20 mins - if not specifying seed and query..
  - Seed - set to 3 because? - Territories to be included in group 1 for DiffGeneExp
    - If I say 3 - does it only use territory 3 for comparison or does it compare 3 different territories?
  - Query - Territories to be included in group 2 of DGE.


seed against query - pick territories manually!! based on the territory plot
  - eg inner vs outer part of metastasis
  - outer part vs normal
  - outer part vs other metastasis
  
Output:
 - p.value,
 - logFC,
 - seedPct (percentage of cells containing gene in first group),
 - queryPct (percentage of cells containing gene in second group),
 - seedTerritory (territory used as group 1)
 - queryTerritory (territory used as group 2)
  
TerritoryMorphing
 - 

```{r}
territoryPlot(vesMorph,cex=10,cex.pt=pt.sz)

# Specify territories for comparison, seed vs query
vs <- "MetVsMetastasis"
s <- c(8,7)
q <- c(2,3)

vesMark <- extractMarkers(vesTerr,method=st.meth, seed=s, query=q, pval=0.05)

vesMorph <- territoryMorphing(vesMark, c(5,6))

#vesTerrLay <- layerTerritory(vesMorph, c(28,34,32),morphologyFactor=0,layerDepth=5)

param <- paste(param, paste0("markerMethod", st.meth), sep="_")
print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))
```





## Pick top marker bins

```{r}
mark <- vesMark@DEG$DEG_Territory_Trial_1
m.df <- as.data.frame(mark)
#padj <- m.df[order(m.df$p.value.adj),]
padj <- slice_min(m.df, order_by=m.df$p.value.adj, n=50)
#logfc.min <- slice_min(m.df, order_by=m.df$logFC, prop=0.1)
#logfc.max <- slice_max(m.df, order_by=m.df$logFC, prop=0.1)
logfc.min <- slice_min(padj, order_by=padj$logFC, n=5)
logfc.max <- slice_max(padj, order_by=padj$logFC, n=5)
int.bins <- bind_rows(list(logfc.min, logfc.max))


```


## Plot marker bins
Investigate individual marker bins over the whole tissue slice.
Are some bin patterns correlating with the territory regions?

The beads are overlapping - change cex size from default 10 to less - or change alpha to change opacity/transparency

```{r}
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerBins", vs, alias, param, "layer-TRUE", ".pdf", sep="_")
pdf(fname)
layout(matrix(c(1,3,2,4), 2, 2))
for(i in seq_along(int.bins$genes)){
  print(viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=TRUE,norm=F,cex=0.5))
  #p <- as.cimg(viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=F,norm=F,cex=2))
  #print(p)
}
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.pathDev, fname))




```







```{r}


mark <- ves@DEG$DEG_Territory_Trial_1
m.df <- as.data.frame(mark)
#padj <- m.df[order(m.df$p.value.adj),]
padj <- slice_min(m.df, order_by=m.df$p.value.adj, prop=0.1)
#logfc.min <- slice_min(m.df, order_by=m.df$logFC, prop=0.1)
#logfc.max <- slice_max(m.df, order_by=m.df$logFC, prop=0.1)
logfc.min <- slice_min(padj, order_by=padj$logFC, prop=0.3)
logfc.max <- slice_max(padj, order_by=padj$logFC, prop=0.3)
int.bins <- bind_rows(list(logfc.min, logfc.max))

for(i in seq_along(int.bins$genes)){
  print(viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=F,norm=F,cex=2))
}

```











```{r}
p <- list()
for(i in seq(4)){
  p[[i]] <- viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=F,norm=F)
}
grid.arrange(p[[1]], p[[2]], nrow = 2)
```

## Where are the bins?
In which chromosomes and where can we find the marker bins of interest. 


```{r}

bin.file <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/mm10_1Mb_bins.txt", header = T)
chr <- data_frame()
for(i in seq_along(int.bins$genes)){
  x <- as.list(subset(bin.file, bin_ind==int.bins$genes[[i]]))
  chr <- bind_rows(chr, x)
}

chr <- chr[order(chr$bin_start),]
chr <- chr[order(chr$chr_ind),]

print(int.bins)
print(chr)
```



