---
title: "SlideSeqDNA_Vesalius_Dev"
output: html_document
---


## Settings - Parameters etc

Specify dataset + Normalisation + dimensionality reduction method and other parameters

For 

```{r}
# SlideSeq DNA datasets:
#alias <- "human_colon_cancer_4_dna_200114_13"
#alias <- "human_colon_cancer_3_dna_191204_19"
alias <- "mouse_liver_met_2_dna_200114_10"



norm <- "TFIDF"             # Normalization method: "log","SCT","TFIDF","raw"
meth <- "LSI"               # Dimensionality reduction method: "PCA","PCA_L","UMAP","LSI","LSI_UMAP"
PCs <- 12                   # Number of Principal Components
plotPC <- 8                 # Number of PC plots to export. 4 plots per pdf page

min.count <- 1
max.count <- 6

k <- 20
knn.type <- "Duplicate"     # KNN method: "Unique", "Duplicate"

cores <- 5

colD <- 3                   # Column Depth
smIter <- 10                # Smoothing Iterations
imSegDims <- 4              # Image Segmentation Dimensions



param <- paste(paste0(norm, "Norm"), paste0(meth, PCs), sep = "_")

in.path <- "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/"
out.path <- "~/Thesis_SpatialGenomics/R_Output/"
out.pathPC <- "~/Thesis_SpatialGenomics/R_Output/PC_images/"

```

In paper they do k 50 for smoothing the beads before PCA (mouse liver)

## Load packages needed

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/")
knitr::opts_chunk$set(echo = TRUE)
library(imagerExtra)
library(Seurat)
library(ggplot2)
library(patchwork)
library(viridis)
library(RColorBrewer)
library(dplyr)
library(tvR)
library(Signac)
library(sp)
library(spdep)
library(grid)
library(Matrix)
library(MatrixExtra)


files <- list.files("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R",pattern = ".R")
files <- paste0("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R/",files)
for(i in seq_along(files)){
  source(files[i])
}

# Standard libraries - just in case you are a command line fiend
library(utils)
library(stats)
library(graphics)
library(grDevices)
### Set seed
set.seed(1)
```



## Read data from files
Vectors:
barcodes    -     holds the barcode sequences
sparse      -     Sparse count matrix, dgCMatrix

```{r}
counts <- paste0(in.path, alias, ".sparse_counts_1Mb.txt")
bead <- ReadSlideSeq(paste0(in.path, alias, ".bead_locations.csv"))

barcodes <- rownames(bead@coordinates)

sep = "\t" #separator used in this count matrix
ctmp <- read.table(counts, header = TRUE, sep=sep ) %>% na.exclude()

sparse <- sparseMatrix(i= ctmp[,2],j=ctmp[,1],x=ctmp[,3]) 

# Set point size for plotting the UMIs - adjusted for knn later
pt.sz <- 0.25
```


## Genomic Bins Human/Mouse:
Each bin index corresponds to 1 million bases in a specific location on a chromosome.
bin size: 1 million bases, index 1 = chromosome 1 from base 1 -> 1.000.000
Keep only the bin index numbers (1-3114. ~3.1bn bases)

Human: 3114 bins human bins. Chr2 starts at bin 251
Mouse: 2738 bins total 

```{r}

if(grepl("human",alias)){
  bins <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/hg19_1Mb_bins.txt", header = T)
  bins <- bins$bin_ind
} else if(grepl("mouse",alias)){
  bins <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/mm10_1Mb_bins.txt", header = T)
  bins <- bins$bin_ind
}
```



## Name column/rows in sparse matrix
Assign bin index numbers to identify the rows. 
Assign the barcodes to identify the columns

```{r}
colnames(sparse) <- barcodes
rownames(sparse) <- bins
```

#Remove outlier counts
Expected CNA varies between 
High outlier counts might represent mitochondrial DNA bins


```{r}

if(exists("min.count")){
  sparse <- filterSparse(sparse, function(x) x >= min.count)
}
if(exists("max.count")){
  sparse <- filterSparse(sparse, function(x) x <= max.count)
  miax <- paste0("counts", min.count, "-", max.count)
  param <- paste(miax, param, sep="_")
}


```


## Vector holders - Checkpoint

```{r}
df1 <- as.matrix(sparse) 
bead1 <- bead
param1 <- param

df <- df1
bead <- bead1
param <- param1
```


## Run KNN
- Decide between only using each barcode once (KNN.DNA.Unique function)
  or having many duplicate groups of barcodes, each barcode gets a group of k nearest neighbors (KNN.DNA function)


```{r}
if(exists("k")){
  if(knn.type == "Unique"){
    ls <- KNN.DNA.Unique(bead, df, k)
  }else{
    ls <- KNN.DNA(bead, df, k)
  }
  param <- paste(param, paste0("knn", knn.type, k), sep="_")
  sparse <- ls[[1]] #Retrieve sparse matrix output from knn function
  bead <- ls[[2]] #Retrieve Bead info (barcodes and coordinates) in SlideSeq assay format compatible with Seurat/Vesalius object

  pt.sz <- k/15 #Adjusting point size for plotting
  rm(ls)
}

bead <- cbind(rownames(bead), bead[[1]], bead[[2]])
colnames(bead) <- c("barcodes","xcoord","ycoord")
bead <- as.data.frame(bead)
bead <- transform(bead, xcoord = as.numeric(xcoord), 
               ycoord = as.numeric(ycoord))

sparse.df <- as.data.frame(sparse)
sparse.spmx <- Matrix(sparse, sparse = TRUE)

```

## Create Vesalius Object - Milestone
Based on the counts and bead coordinates

Beads in data.frame format - coordinates numerical
Counts either dataframe or sparse matrix.

# Dimensionality Reduction
Latent space
PCA / UMAP / tSNE (/LSI)

```{r}
#sparse <- sparse.df
sparse <- sparse.spmx

vesalius <- buildVesaliusObject(bead,sparse)
ves <- buildVesaliusEmbeddings(vesalius, method = meth, 
  norm = norm, pcs = PCs, tensorResolution = 0.1, cores = cores)

```

## Plotting vesalius images of the PC or LSI embeddings

Imageplot with chosen number of dimensions for the different PCs

```{r}
cat( paste(Sys.time()," Plotting top", plotPC, "PCs and exporting to pdf.\n File located in:", out.pathPC, "\n"))
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), alias, param, "VesDev_PCimages.pdf", sep="_")
pdf(fname)
layout(matrix(c(1,3,2,4), 2, 2))
for(i in seq(plotPC)){
  cat(" Hi ", i, "\n")
  imagePlot(ves, dims = i)
}
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.pathPC, fname))
```


## Image processing:
imageSegmentation - not sure about the dims here - is that PC dimensions?
  - Embedding "last", last what?

```{r}
#vesImSeg <- imageSegmentation(ves, dims = imSegDims, colDepth = colD, smoothIter = smIter, smoothType = "iso", sigma = 2, embedding = "PCA")
vesImSeg <- imageSegmentation(ves,dims=1:30, colDepth = 5,smoothIter=5,smoothType="iso",sigma =2,embedding = "PCA")

vesTerr <- isolateTerritories(vesImSeg, trial = "last")

vesMark <- extractMarkers(vesTerr, seed = 7)

vesMorph <- territoryMorphing(vesMark, c(5,6))

vesTerrLay <- layerTerritory(vesMorph, c(28,34,32),morphologyFactor=0,layerDepth=5)

#viewGeneExpression(ves, gene= "Cryba4", as.layer=T,norm=F)
```




## Imsplit for grayscale
gray.cimg.plot <- function(image, step){
  ss.split <- subset(image.array, select=c(x, y, cc, value))
  split.im <- as.cimg(ss.split)
  split.image <- imsplit(split.im, "cc", 3)
}


## Publish plots in pdf
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), alias, param, "images_VesDev.pdf", sep="_")
pdf(fname)
#layout(matrix(c(1,3,2,4), 2, 2))
#plot(split[[1]], main=c('Original ImageArray, 1st colour channel'))

#cimg.plot(image.array, 'Original')
#cimg.plot(image, 'Isolated Territories')


territoryPlot(image,cex=10,cex.pt=pt.sz)

dev.off()

file.rename(from = paste0(in.path, fname), to = paste0(out.path, fname))

```{r}

```

