---
title: "SlideSeqRNA_Vesalius_Dev"
output: html_document
---


## Settings - Parameters etc

Specify dataset + Normalisation + dimensionality reduction method and other parameters

Available Slide-RNA-Seq datasets
human_colon_cancer_4_rna_200102_06
mouse_liver_met_2_rna_201002_04

```{r}
start.time <- Sys.time()
# SlideSeq RNA datasets:

alias <- "mouse_liver_met_2"
al.num <- "_rna_201002_04"

norm <- "SCT"             # Normalization method: "log","SCT","TFIDF","raw"
meth <- "PCA"               # Dimensionality reduction method: "PCA","PCA_L","UMAP","LSI","LSI_UMAP"
PCs <- 30                   # Number of Embeddings / dimensions / Principal Components / "LSI topics"
plotPC <- 8                 # Number of PC plots to export. 


cores <- 5

colD <- 6                   # Column Depth (3-6)
smIter <- 10                # Smoothing Iterations
#imSegDims <- 4             # Image Segmentation Dimensions
st.meth <- "chisq"          # Statistical method for extracting marker genes - "wilcox", "t.test", "chisq"


in.path <- "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/"
out.path <- paste0("~/Thesis_SpatialGenomics/R_Output/VesaliusDev/RNA/", alias, "/")


```


## Load packages needed

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/")
knitr::opts_chunk$set(echo = TRUE)
library(imagerExtra)
library(Seurat)
library(ggplot2)
library(patchwork)
library(viridis)
library(RColorBrewer)
library(dplyr)
library(tvR)
library(Signac)
library(sp)
library(spdep)
library(grid)
library(gridExtra)
library(Matrix)
library(MatrixExtra)


files <- list.files("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R",pattern = ".R")
files <- paste0("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R/",files)
for(i in seq_along(files)){
  source(files[i])
}

# Standard libraries - just in case you are a command line fiend
library(utils)
library(stats)
library(graphics)
library(grDevices)
### Set seed
set.seed(1)
```



## Read data from files
Vectors:
barcodes    -     holds the barcode sequences
sparse      -     Sparse count matrix, dgCMatrix

```{r}
counts <- paste0(in.path, alias, al.num, ".sparse_expression.txt")
bead <- ReadSlideSeq(paste0(in.path, alias, al.num, ".bead_locations.csv"))

barcodes <- rownames(bead@coordinates)

sep = "," #separator used in this count matrix
ctmp <- read.table(counts, header = TRUE, sep=sep ) %>% na.exclude()
#rownames(ctmp) <- ctmp[,1] #Non-Unique values - are these the genes/features?
#ctmp <- ctmp[,-1]

sparse <- sparseMatrix(i= ctmp[,1],j=ctmp[,2],x=ctmp[,3])


# Set point size for plotting the UMIs - adjusted for knn later
pt.sz <- 0.25
```


## Parameters
Keep track of parameters in param vector for file naming etc.

```{r}
param <- paste(paste0(norm, "Norm"), paste0(meth, PCs), sep = "_")
```



## Features
Assign either the mouse or human genes as rownames of count matrix

```{r}
if(grepl("human",alias)){
  genes <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/human_genes.txt", header = T)
} else {
  genes <- read.table("~/Thesis_SpatialGenomics/SlideSeq/GroupSlideSeq/mouse_genes.txt", header = T)
}
```



## Name column/rows in sparse matrix
Assign gene names to identify the rows. 
Assign the barcodes to identify the columns

```{r}
colnames(sparse) <- barcodes
rownames(sparse) <- genes[,1]
bead <- bead@coordinates
bead <- cbind(rownames(bead), bead[[1]], bead[[2]])
colnames(bead) <- c("barcodes","xcoord","ycoord")
```



## Vector holders - Checkpoint

```{r}
df1 <- as.matrix(sparse) 
bead1 <- bead
param1 <- param

df <- df1
bead <- bead1
param <- param1
```



## Create Vesalius Object - Milestone
Based on the counts and bead coordinates

Beads in data.frame format - coordinates numerical
Counts either dataframe or sparse matrix.

# Dimensionality Reduction
Latent space
PCA / UMAP / tSNE (/LSI)

set filterThreshold to remove outlier beads (0.99)
```{r}

vesalius <- buildVesaliusObject(bead,df)
ves <- buildVesaliusEmbeddings(vesalius, method = meth, 
  norm = norm, pcs = PCs, tensorResolution = 0.1, cores = cores, filterThreshold =0.99)

```

## Plotting vesalius images of the PC or LSI embeddings

Imageplot with chosen number of dimensions (for the different PCs)

```{r}
cat( paste(Sys.time()," Plotting top", plotPC, "PCs and exporting to pdf.\n File located in:", out.path, "\n"))
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "Embeddings", alias, param, "VesDev.pdf", sep="_")
pdf(fname)
#layout(matrix(c(1,3,2,4), 2, 2))
for(i in seq(plotPC)){
  print(imagePlot(ves, dims = i))
}
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.path, fname))

print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))
```


## Image processing:
imageSegmentation - dims: number of dimred embeddings
  - Embedding "last": last dimred method used (if you have multiple dimred objects stored in your ves object)
LayerTerritory - not relevant for DNA data?

```{r}
vesImSeg <- imageSegmentation(ves,dims=1:PCs, colDepth = colD, smoothIter=smIter,smoothType="iso",sigma=2,embedding=meth)

vesTerr <- isolateTerritories(vesImSeg, trial = "last")

vesMark <- extractMarkers(vesTerr, seed = 3, method=st.meth)

vesMorph <- territoryMorphing(vesMark, c(5,6))

#vesTerrLay <- layerTerritory(vesMorph, c(28,34,32),morphologyFactor=0,layerDepth=5)

param <- paste(param, paste0("dims1-", PCs), paste0("colD", colD), paste0("Iter", smIter), "markerMethod", st.meth, sep="_")
```



# Export territory plots to pdf
Is there a difference between the territory isolation and the territory morphing?
Also plot the first imagesegmentation

```{r}
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "Territories", alias, paste0(param, ".pdf"), sep="_")
pdf(fname)
layout(matrix(c(1,3,2,4), 2, 2))
territoryPlot(vesTerr,cex=10,cex.pt=pt.sz)
print(imagePlot(vesImSeg))
print(getLastLog(vesTerr))
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.path, fname))
```



```{r}
territoryPlot(vesTerr,cex=10,cex.pt=pt.sz)
param2 <- param

```
## Stop in Pipeline
Change parameters

```{r}
change territories and info in the next chunk!!
```


## Extract marker bins and morph territories
ExtractMarkers - DGE between selected territories - identify marker bins
  - Takes like 20 mins - if not specifying seed and query..
      - More territories = more time?
  - Seed - set to 3 because? - Territories to be included in group 1 for DiffGeneExp
    - If I say 3 - does it only use territory 3 for comparison or does it compare 3 different territories?
  - Query - Territories to be included in group 2 of DGE.


seed against query - pick territories manually!! based on the territory plot
  - eg inner vs outer part of metastasis
  - outer part vs normal
  - outer part vs other metastasis
  
Output:
 - p.value,
 - logFC,
 - seedPct (percentage of cells containing gene in first group),
 - queryPct (percentage of cells containing gene in second group),
 - seedTerritory (territory used as group 1)
 - queryTerritory (territory used as group 2)
  
TerritoryMorphing
 - Territory morphing can manipulate territories by growing, shrinking, filling, and cleaning territories.
 - Growing = Positive integers - Territory will be dilated by x number of pixels
 - Shrinking = Negative integers - Territory will be contracted by x number of pixels
 - Filling = grow followed by shrink.
 - Cleaning = shrink followed by grow.
 - Morphology factor - integer or vector of integers describing growth and/or shrink extent.

```{r}


# Specify territories for comparison, seed vs query
vs <- "CloneBVsCloneA"
s <- c(14,13)           #c(8,7) or NULL
q <- c(5,7)             #c(2,3) or NULL
pval <- 0.05

vesMark <- extractMarkers(vesTerr,method=st.meth, seed=s, query=q, pval=pval)

vesMorph <- territoryMorphing(vesMark, c(1,10))

#vesTerrLay <- layerTerritory(vesMorph, c(28,34,32),morphologyFactor=0,layerDepth=5)

param <- paste(param2, paste0("markerMethod", st.meth), sep="_")
print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))
```





## Pick top marker bins
 - Based on lowest adjusted p value and highest, furthest from 0, Log fold change
 - With RNA we only get around 25 Genes - no need to subset..

```{r}
mark <- vesMark@DEG$DEG_Territory_Trial_1
#m.df <- as.data.frame(mark)
#padj <- slice_min(m.df, order_by=m.df$p.value.adj, n=100)
#logfc.min <- slice_min(padj, order_by=padj$logFC, n=5)
#logfc.max <- slice_max(padj, order_by=padj$logFC, n=5)
#int.genes <- bind_rows(list(logfc.min, logfc.max))
int.genes <- as.data.frame(mark)
int.genes <- int.genes[order(int.genes$p.value.adj),]
```

population[order(population$age),]


## Plot marker bins - export to pdf
Investigate individual marker bins over the whole tissue slice.
Are some bin patterns correlating with the territory regions?

The beads are overlapping - change cex size from default 10 to less - or change alpha to change opacity/transparency






viewGeneExpression
  - why vesMorph as input? better with territory?
  - as.layer - should it show marker bin's expression/distribution per bead or overall per territory? - TRUE / FALSE
  





```{r}
layer <- "FALSE"           
p <- list()
for(i in seq(int.genes$genes)){
  p[[i]] <- viewGeneExpression(vesMorph, gene=int.genes$genes[[i]], as.layer=layer,norm=F, cex=0.1, territory1=s, territory2=q)
}

setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerGenes", vs, alias, param, paste0("layer", layer), ".pdf", sep="_")
pdf(fname)
grid.arrange(p[[1]], p[[2]], p[[3]], p[[4]], nrow = 2)
grid.arrange(p[[5]], p[[6]], p[[7]], p[[8]], nrow = 2)
grid.arrange(p[[9]], p[[10]], nrow = 2)
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.path, fname))
```

## Another bin plotting version: (One plot per pdf page)
setwd(in.path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerGenes", vs, alias, param, paste0("layer-", layer), ".pdf", sep="_")
pdf(fname)
for(i in seq_along(int.genes$genes)){
  print(viewGeneExpression(vesMorph, gene=int.genes$genes[[i]], as.layer=layer,norm=F,cex=0.5))
}
dev.off()
file.rename(from = paste0(in.path, fname), to = paste0(out.path, fname))

## Plot bins here in markdown:
for(i in seq_along(int.genes$genes)){
  print(viewGeneExpression(vesMorph, gene=int.genes$genes[[i]], as.layer=F,norm=F,cex=2))
}



Here I need to investigate the genes... How?!
 


```{r}

fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerBins", vs, alias, param, "table.csv", sep="_")
write.csv(int.genes, file=paste0(out.path,fname))

print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))

```



DNA:
## Chromosomal location
In which chromosomes and where can we find the marker bins? 
 - bin.all (vector with table from either human or mouse chromosomes and the bin indexes)
 - find chromosome location for each bin index
 - keep only bin index, start, stop, and chromosome number 
 - merge the location information with the statistical output for each bin
 - export table to csv
 
chr <- data_frame()
for(i in seq_along(padj$genes)){
  x <- as.list(subset(genes[,1], bin_ind==padj$genes[[i]]))
  chr <- bind_rows(chr, x)
}
chr <- chr[c("bin_start", "bin_end", "chr_ind", "bin_ind")]

names(padj)[names(padj) == 'genes'] <- 'bin_ind'
bin.chr <- merge(chr, padj)
bin.chr <- bin.chr[order(bin.chr$bin_start),]
bin.chr <- bin.chr[order(bin.chr$chr_ind),]

fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerBins", vs, alias, param, "table.csv", sep="_")
write.csv(bin.chr, file=paste0(out.path,fname))

print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))


## Check the log of your current vesalius object
getLastLog(vesTerr)