---
title: "PostVesaliusTerritoriesDNA"
author: "Heidi Ottesen"
date: '2022-07-06'
output: html_document
---

For investigating layers within a clone
and also for comparing marker bins with the marker genes from the same RNA layers


## Load packages and functions

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
  
library(RColorBrewer)
library(ggplot2)
library(stats)

#files <- paste0("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R/",list.files("~/Thesis_SpatialGenomics/VesaliusDev_2022/Vesalius/R",pattern = ".R"))
## If on macbook:
files <- paste0("~/Thesis/Vesalius/R/",list.files("~/Thesis/Vesalius/R",pattern = ".R"))
for(i in seq_along(files)){
  source(files[i])
}
```

## Load vesalius Territory object
and have a look

RNA:
Object_Territories_mouse_liver_met_2_SCTNorm_PCA30_dims1-30_colD6_Iter10_markerMethod_chisq

DNA:
Object_DNA_Territories_mouse_liver_met_2_counts0-6_TFIDFNorm_LSI10_knnDuplicate25_dims1-10_colD6_Iter10

Object_DNA_Territories_mouse_liver_met_2_counts0-6_logNorm_PCA10_knnDuplicate25_dims1-10_colD6_Iter10
Object_DNA_Territories_mouse_liver_met_2_counts0-6_logNorm_PCA10_knnDuplicate5_dims1-10_colD6_Iter10
Object_DNA_Territories_mouse_liver_met_2_counts0-6_logNorm_PCA10_knnDuplicate10_dims1-10_colD6_Iter10
Object_DNA_Territories_mouse_liver_met_2_counts0-6_logNorm_PCA10_knnDuplicate50_dims1-10_colD6_Iter10

Object_DNA_Territories_mouse_liver_met_2_ChrSummed_counts0-6_logNorm_PCA10_dims1-10_colD6_Iter10
Object_DNA_Territories_mouse_liver_met_2_ChrSummed_counts0-6_logNorm_PCA10_knnDuplicate5_dims1-10_colD6_Iter10
Object_DNA_Territories_mouse_liver_met_2_ChrSummed_counts0-6_logNorm_PCA10_knnDuplicate25_dims1-10_colD6_Iter10


```{r}
fname <- "Object_DNA_Territories_mouse_liver_met_2_ChrSummed_counts0-6_logNorm_PCA10_knnDuplicate25_dims1-10_colD6_Iter10"


#ref.path <- "~/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/"
#path <- "~/Thesis_SpatialGenomics/R_Output/VesaliusDev/RNA/mouse_liver_met_2/"
##MacBook:
path <- "~/Thesis/R_Output/VesaliusDev/mouse_liver_met_2/"
ref.path <- "~/Thesis/Thesis_SpatialGenomics/SlideSeq/SlideSeq_ves_test/"

load(file=paste0(path, fname, ".rd"))
territoryPlot(vesTerrDNA,cex=10,cex.pt=0.25)
#territoryPlot(vesTerr,cex=10,cex.pt=0.25, split=TRUE)


Pick territories in next block
```


## Specify clone 
Select territories and statistical method for clonal comparisons.

seed against query - pick territories manually!! based on the territory plot
  - eg inner vs outer part of metastasis    "CloneAInnerVsOuter", "CloneBInnerVsOuter"
  - outer part vs normal(NULL)                    
  - outer part vs other metastasis
  - metastasis vs metastasis                "CloneBVsCloneA"
  - Both metastases vs normal               "CloneBandAvsNormal"
  
## Sublayers:
Specify which territories to investigate further and desired number of sublayers
  
```{r}

vs <- "CloneBvsCloneA"
s <- c(13, 14)           
q <- 2

st.meth <- "wilcox"
pval <- 0.05

# Sublayers
sublayers <- 5
subS <- c(13, 14)


```


## Extract marker bins and morph territories
ExtractMarkers - DGE between selected territories - identify marker bins
  - Takes like 20 mins - if not specifying seed and query..
      - More territories = more time?
  - Seed - set to 3 because? - Territories to be included in group 1 for DiffGeneExp
    - If I say 3 - does it only use territory 3 for comparison or does it compare 3 different territories?
  - Query - Territories to be included in group 2 of DGE.
  
  Output: p.value,- logFC, - seedPct (percentage of cells containing gene in first group), - queryPct (percentage of cells containing gene in second group), - seedTerritory (territory used as group 1) - queryTerritory (territory used as group 2)
  
## Identify sublayers
For further investigating the underlying gene expression

layerTerritory() - 
  - morphologyFactor - shrink and expand in pixels (dilation/Erosion) - to remove tiny outsider blobs - doesn't work atm, keep at 0
  - choose territories of interest to further layer
  layerDepth - number of sub-territories/layers you would like (for investigating gradual change)

  

  
TerritoryMorphing
  - c(1,10) - Shrinking/Growing factor??!!! - Nope territories
 - Territory morphing can manipulate territories by growing, shrinking, filling, and cleaning territories.
 - Growing = Positive integers - Territory will be dilated by x number of pixels
 - Shrinking = Negative integers - Territory will be contracted by x number of pixels
 - Filling = grow followed by shrink.
 - Cleaning = shrink followed by grow.
 - Morphology factor - integer or vector of integers describing growth and/or shrink extent.




```{r}

vesMark <- extractMarkers(vesTerrDNA, method=st.meth, seed=s, query=q, pval=pval)
#vesTerrLay <- layerTerritory(vesMark, c(s,q),morphologyFactor=0,layerDepth=5)

````




## Pick top marker bins
 - Based on lowest adjusted p value and highest, furthest from 0, Log fold change

```{r}
int.bins <- vesMark@DEG$DEG_Territory_Trial_1
#mark <- as.data.frame(mark)

length(int.bins$genes)
Appropriate number of bins?

padj <- slice_min(int.bins, order_by=int.bins$p.value.adj, n=100)
int.bins <- padj

length(int.bins$genes)

```


Optional further subsetting:
(Otherwise skip)


```{r}
logfc.min <- slice_min(int.bins, order_by=int.bins$logFC, n=20)
logfc.max <- slice_max(int.bins, order_by=int.bins$logFC, n=20)
int.bins <- bind_rows(list(logfc.min, logfc.max))
length(int.bins$genes)
```

## Plot marker bins - export to pdf
Investigate individual marker bins over the whole tissue slice.
Are some bin patterns correlating with the territory regions?

The beads are overlapping - change cex size from default 10 to less - or change alpha to change opacity/transparency




viewGeneExpression
  - Use object from extractMarkers()
  - as.layer - should it show marker bin's counts/distribution per bead or overall per territory? - TRUE / FALSE
  



```{r}

layer <- "FALSE" 
for(i in seq(int.bins$genes)){
  f <- paste0(path, "BinCountLayer/", vs, "/", int.bins$genes[[i]], "_", vs, "_", fname, ".png")
  png(f) 
  print(viewGeneExpression(vesMark, gene=int.bins$genes[[i]], as.layer=layer, norm=F, cex=0.15, territory1=s, territory2=q))
  dev.off()
}


```

## Another bin plotting version: (One plot per pdf page - Huge file!)
setwd(path)
fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerBins", vs, alias, param, paste0("layer-", layer), ".pdf", sep="_")
pdf(fname)
for(i in seq_along(int.bins$genes)){
  print(viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=layer,norm=F,cex=0.5))
}
dev.off()
file.rename(from = paste0(path, fname), to = paste0(out.path, fname))

## Plot bins here in markdown:
for(i in seq_along(int.bins$genes)){
  print(viewGeneExpression(vesMorph, gene=int.bins$genes[[i]], as.layer=F,norm=F,cex=2))
}



## Chromosomal location


Reference bins - Humans or mouse?
Assemblies:
  					® Human Genome used for the hg19 bins: GRCh37
						◊    https://www.ncbi.nlm.nih.gov/assembly/GCF_000001405.13/#/st
					® Mouse genome mm10 bins: GRCm38
						◊ https://www.ncbi.nlm.nih.gov/assembly/GCF_000001635.20/#/st
          
```{r}
if(grepl("human",fname)){
  bin.all <- read.table(paste0(ref.path,"hg19bins.txt"), header = T)
  bins <- bin.all$bin_ind
} else if(grepl("mouse",fname)){
  bin.all <- read.table(paste0(ref.path,"mm10bins.txt"), header = T)
  bins <- bin.all$bin_ind
}
```
 In which chromosomes and where can we find the marker bins? 
 - bin.all (vector with table from either human or mouse chromosomes and the bin indexes)
 - find chromosome location for each bin index
 - keep only bin index, start, stop, and chromosome number 
 - merge the location information with the statistical output for each bin
 - export table to csv


## Get bin location info w Function:
```{r}
if(grepl("Chromosome",fname)){
  bin.chr <- binLocationsChr(padj, bin.all)
}else{
  bin.chr <- binLocations(int.bins, bin.all)
}


write.csv(bin.chr, file=paste0(path,"MarkerBins_", fname, ".csv"))


```

chr <- data_frame()
for(i in seq_along(padj$genes)){
  x <- as.list(subset(bin.all, bin_ind==padj$genes[[i]]))
  chr <- bind_rows(chr, x)
}
chr <- chr[c("bin_start", "bin_end", "chr_ind", "bin_ind")]

names(padj)[names(padj) == 'genes'] <- 'bin_ind'
bin.chr <- merge(chr, padj)
bin.chr <- bin.chr[order(bin.chr$bin_start),]
bin.chr <- bin.chr[order(bin.chr$chr_ind),]

fname <- paste(format(Sys.time(), "%Y%m%d_%H%M%S"), "MarkerBins", vs, alias, param, "table.csv", sep="_")
write.csv(bin.chr, file=paste0(out.path,fname))

print(paste("Runtime so far: ", difftime(Sys.time(), start.time, units="mins"), "min"))





## Get genes from top bins:
```{r}
genesDNA <- genesBioMart(bin.chr, fname)
genesDNA <- genesDNA[!grepl('^predicted', genesDNA$description), ] #Remove unstudied predicted genes

write.csv(genesDNA, file=paste0(path,paste0("Genes_", fname, ".csv")))
#save(genesDNA, file=paste0(path, "Genes_from_DNA_Bins_", fname, ".rd"))

```




## Check the log of your current vesalius object
getLastLog(vesTerr)



  

